src/js/midi/math.js

```diff
@@ -1,46 +1,48 @@
-const add = R.curry((a, b) => Decimal.add(a, b))
+const add = (a, b) => Decimal.add(a, b)
 
-const subtract = R.curry((a, b) => Decimal.sub(a, b))
+const subtract = (a, b) => Decimal.sub(a, b)
 
-const multiply = R.curry((a, b) => Decimal.mul(a, b))
+const multiply = (a, b) => Decimal.mul(a, b)
 
-const divide = R.curry((a, b) => Decimal.div(a, b))
+const divide = (a, b) => Decimal.div(a, b)
 
-const pow = R.curry((a, b) => Decimal.pow(a, b))
+const pow = (a, b) => Decimal.pow(a, b)
 
-const log = R.curry((a, b) => Decimal.log(a, b))
+const log = (a, b) => Decimal.log(a, b)
 
 const floor = (a) => Decimal.floor(a)
 
-const moveNUnits = R.curryN(4, (ratioOfSymmetry, divisionsPerRatio, n, frequency) =>
-  R.compose(multiply(frequency), pow(ratioOfSymmetry), divide(n))(divisionsPerRatio)
-)
+const moveNUnits = (ratioOfSymmetry, divisionsPerRatio, n, frequency) => {
+  // return frequency * ratioOfSymmetry ** (n / divisionsPerRatio)
+  return multiply(frequency, pow(ratioOfSymmetry, divide(n, divisionsPerRatio)))
+}
 
-const getDistanceInUnits = R.curryN(
-  4,
-  (ratioOfSymmetry, divisionsPerRatio, frequency2, frequency1) =>
-    R.compose(
-      multiply(divisionsPerRatio),
-      log(R.__, ratioOfSymmetry),
-      divide(frequency2)
-    )(frequency1)
-)
+const getDistanceInUnits = (ratioOfSymmetry, divisionsPerRatio, freq2, freq1) => {
+  // return divisionsPerRatio * Math.log(freq2 / freq1, ratioOfSymmetry)
+  return multiply(divisionsPerRatio, log(divide(freq2, freq1), ratioOfSymmetry))
+}
 
-const moveNSemitones = moveNUnits(octaveRatio, semitonesPerOctave)
-const getDistanceInSemitones = getDistanceInUnits(octaveRatio, semitonesPerOctave)
+const moveNSemitones = (n, frequency) => {
+  return moveNUnits(octaveRatio, semitonesPerOctave, n, frequency)
+}
+
+const getDistanceInSemitones = (freq2, freq1) => {
+  return getDistanceInUnits(octaveRatio, semitonesPerOctave, freq2, freq1)
+}
 
 const bendingRatio = moveNSemitones(maxBendingDistanceInSemitones, 1)
 
-const getBendingDistance = getDistanceInUnits(bendingRatio, pitchBendMax)
+const getBendingDistance = (freq2, freq1) => {
+  return getDistanceInUnits(bendingRatio, pitchBendMax, freq2, freq1)
+}
 
-const getNoteFrequency = R.compose(
-  moveNSemitones(R.__, referenceNote.frequency),
-  subtract(R.__, referenceNote.id),
-  R.clamp(keyIdMin, keyIdMax)
-)
+const getNoteFrequency = (midinote) => {
+  return moveNSemitones(
+    subtract(R.clamp(0, 127, midinote), referenceNote.id),
+    referenceNote.frequency
+  )
+}
 
-const getNoteId = R.compose(
-  floor,
-  add(R.__, referenceNote.id),
-  getDistanceInSemitones(R.__, referenceNote.frequency)
-)
+const getNoteId = (frequency) => {
+  return floor(add(getDistanceInSemitones(frequency, referenceNote.frequency), referenceNote.id))
```

Comment:
Doesn't this imply that we're always bending upwards? Why not `round` instead of `floor`?

Reply:
Because it makes other calculations easier, if I don't have to deal with checking the closest note both upwards and downwards. Using floor means a frequency is always made up of `basenote + pitchbending`. On the other hand using round would mean, that a note can be either `basenote + pitchbending` or `basenote - pitchbending`. All other calculations assume pitchbending to be positive.